File: service/javadsl/api/src/main/java/com/lightbend/lagom/javadsl/api/deser/MessageSerializers.java
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.javadsl.api.deser;

import akka.util.ByteString;
import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.transport.MessageProtocol;
import com.lightbend.lagom.javadsl.api.transport.NotAcceptable;
import com.lightbend.lagom.javadsl.api.transport.UnsupportedMediaType;

import java.nio.charset.Charset;
import java.util.List;
import java.util.Optional;

/**
 * Message serializers.
 */
public class MessageSerializers {

    private MessageSerializers() {}

    /**
     * A serializer for {@link NotUsed} (empty) messages.
     */
    public static final StrictMessageSerializer<NotUsed> NOT_USED = new StrictMessageSerializer<NotUsed>() {
        @Override
        public NegotiatedSerializer<NotUsed, ByteString> serializerForRequest() {
            return aUnit -> ByteString.empty();
        }

        @Override
        public NegotiatedDeserializer<NotUsed, ByteString> deserializer(MessageProtocol messageProtocol) throws SerializationException {
            return bytes -> NotUsed.getInstance();
        }

        @Override
        public NegotiatedSerializer<NotUsed, ByteString> serializerForResponse(List<MessageProtocol> acceptedMessageProtocols) {
            return aUnit -> ByteString.empty();
        }

        @Override
        public boolean isUsed() {
            return false;
        }
    };

    /**
     * A serializer for plain text messages.
     */
    public static final StrictMessageSerializer<String> STRING = new StrictMessageSerializer<String>() {

        MessageProtocol defaultProtocol = new MessageProtocol(Optional.of("text/plain"), Optional.of("utf-8"), Optional.empty());

        class StringSerializer implements NegotiatedSerializer<String, ByteString> {
            private final MessageProtocol protocol;

            StringSerializer(MessageProtocol protocol) {
                this.protocol = protocol;
            }

            @Override
            public MessageProtocol protocol() {
                return protocol;
            }

            @Override
            public ByteString serialize(String s) throws SerializationException {
                return ByteString.fromString(s, protocol.charset().orElse("utf-8"));
            }
        }

        class StringDeserializer implements NegotiatedDeserializer<String, ByteString> {
            private final String charset;

            StringDeserializer(String charset) {
                this.charset = charset;
            }

            @Override
            public String deserialize(ByteString wire) throws DeserializationException {
                return wire.decodeString(charset);
            }
        }

        @Override
        public NegotiatedSerializer<String, ByteString> serializerForRequest() {
            return new StringSerializer(defaultProtocol);
        }

        @Override
        public NegotiatedDeserializer<String, ByteString> deserializer(MessageProtocol protocol) throws UnsupportedMediaType {
            if (protocol.contentType().orElse("text/plain").equals("text/plain")) {
                return new StringDeserializer(protocol.charset().orElse("utf-8"));
            } else {
                throw new UnsupportedMediaType(protocol, defaultProtocol);
            }
        }

        @Override
        public NegotiatedSerializer<String, ByteString> serializerForResponse(List<MessageProtocol> acceptedMessageProtocols) throws NotAcceptable {
            if (acceptedMessageProtocols.isEmpty()) {
                return serializerForRequest();
            } else {
                for (MessageProtocol messageProtocol: acceptedMessageProtocols) {
                    String contentType = messageProtocol.contentType().orElse("text/plain");

                    if (contentType.equals("text/plain") || contentType.equals("text/*") || contentType.equals("*/*") || contentType.equals("*")) {
                        return new StringSerializer(messageProtocol.withContentType("text/plain"));
                    }
                }
                throw new NotAcceptable(acceptedMessageProtocols, defaultProtocol);
            }
        }
    };
}

File: docs/manual/scala/guide/services/code/MessageSerializers.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.scaladsl.services.serializers

package explicitserializers {

  import com.lightbend.lagom.scaladsl.api.Service
  import com.lightbend.lagom.scaladsl.api.ServiceCall

  //#explicit-serializers
  import com.lightbend.lagom.scaladsl.api.deser.MessageSerializer

  trait HelloService extends Service {
    def sayHello: ServiceCall[String, String]

    override def descriptor = {
      import Service._

      named("hello").withCalls(
        call(sayHello)(MessageSerializer.StringMessageSerializer, MessageSerializer.StringMessageSerializer)
      )
    }
  }
  //#explicit-serializers
}

package differentserializers {

  import akka.NotUsed
  import akka.util.ByteString
  import com.lightbend.lagom.scaladsl.api.deser.MessageSerializer
  import com.lightbend.lagom.scaladsl.api.Service
  import com.lightbend.lagom.scaladsl.api.ServiceCall

  //#case-class-two-formats
  import play.api.libs.json._
  import play.api.libs.functional.syntax._

  case class MyMessage(id: String)

  object MyMessage {

    implicit val format: Format[MyMessage] = Json.format
    val alternateFormat: Format[MyMessage] = {
      (__ \ "identifier")
        .format[String]
        .inmap(MyMessage.apply, _.id)
    }
  }
  //#case-class-two-formats

  //#descriptor-two-formats
  trait MyService extends Service {
    def getMessage: ServiceCall[NotUsed, MyMessage]
    def getMessageAlternate: ServiceCall[NotUsed, MyMessage]

    override def descriptor = {
      import Service._

      named("my-service").withCalls(
        call(getMessage),
        call(getMessageAlternate)(
          implicitly[MessageSerializer[NotUsed, ByteString]],
          MessageSerializer.jsValueFormatMessageSerializer(
            implicitly[MessageSerializer[JsValue, ByteString]],
            MyMessage.alternateFormat
          )
        )
      )
    }
  }
  //#descriptor-two-formats
}

package customserializers {

  //#plain-text-serializer
  import akka.util.ByteString
  import com.lightbend.lagom.scaladsl.api.deser.MessageSerializer.NegotiatedSerializer
  import com.lightbend.lagom.scaladsl.api.transport.DeserializationException
  import com.lightbend.lagom.scaladsl.api.transport.MessageProtocol
  import com.lightbend.lagom.scaladsl.api.transport.NotAcceptable
  import com.lightbend.lagom.scaladsl.api.transport.UnsupportedMediaType

  class PlainTextSerializer(val charset: String) extends NegotiatedSerializer[String, ByteString] {
    override val protocol = MessageProtocol(Some("text/plain"), Some(charset))

    def serialize(s: String) = ByteString.fromString(s, charset)
  }
  //#plain-text-serializer

  //#json-text-serializer
  import play.api.libs.json.Json
  import play.api.libs.json.JsString

  class JsonTextSerializer extends NegotiatedSerializer[String, ByteString] {

    override val protocol = MessageProtocol(Some("application/json"))

    def serialize(s: String) =
      ByteString.fromString(Json.stringify(JsString(s)))
  }
  //#json-text-serializer

  //#plain-text-deserializer
  import com.lightbend.lagom.scaladsl.api.deser.MessageSerializer.NegotiatedDeserializer

  class PlainTextDeserializer(val charset: String) extends NegotiatedDeserializer[String, ByteString] {

    def deserialize(bytes: ByteString) =
      bytes.decodeString(charset)
  }
  //#plain-text-deserializer

  //#json-text-deserializer
  import scala.util.control.NonFatal

  class JsonTextDeserializer extends NegotiatedDeserializer[String, ByteString] {

    def deserialize(bytes: ByteString) = {
      try {
        Json.parse(bytes.iterator.asInputStream).as[String]
      } catch {
        case NonFatal(e) => throw DeserializationException(e)
      }
    }
  }
  //#json-text-deserializer

  //#text-serializer
  import com.lightbend.lagom.scaladsl.api.deser.StrictMessageSerializer

  class TextMessageSerializer extends StrictMessageSerializer[String] {
    //#text-serializer

    //#text-serializer-protocols
    override def acceptResponseProtocols = List(
      MessageProtocol(Some("text/plain")),
      MessageProtocol(Some("application/json"))
    )
    //#text-serializer-protocols

    //#text-serializer-request
    def serializerForRequest = new PlainTextSerializer("utf-8")
    //#text-serializer-request

    //#text-deserializer
    def deserializer(protocol: MessageProtocol) = {
      protocol.contentType match {
        case Some("text/plain") | None =>
          new PlainTextDeserializer(protocol.charset.getOrElse("utf-8"))
        case Some("application/json") =>
          new JsonTextDeserializer
        case _ =>
          throw UnsupportedMediaType(protocol, MessageProtocol(Some("text/plain")))
      }
    }
    //#text-deserializer

    //#text-serializer-response
    import scala.collection.immutable

    def serializerForResponse(accepted: immutable.Seq[MessageProtocol]) = {
      accepted match {
        case Nil => new PlainTextSerializer("utf-8")
        case protocols =>
          protocols
            .collectFirst {
              case MessageProtocol(Some("text/plain" | "text/*" | "*/*" | "*"), charset, _) =>
                new PlainTextSerializer(charset.getOrElse("utf-8"))
              case MessageProtocol(Some("application/json"), _, _) =>
                new JsonTextSerializer
            }
            .getOrElse {
              throw NotAcceptable(accepted, MessageProtocol(Some("text/plain")))
            }
      }
    }
    //#text-serializer-response
  }

}

package protobuf {

  import java.io.InputStream
  import java.io.OutputStream

  // Not real protobuf generated class...
  object Order {
    def parseFrom(is: InputStream): Order = null
  }

  class Order {
    def writeTo(os: OutputStream) {}
  }

  //#protobuf
  import akka.util.ByteString
  import com.lightbend.lagom.scaladsl.api.deser.MessageSerializer.NegotiatedDeserializer
  import com.lightbend.lagom.scaladsl.api.deser.MessageSerializer.NegotiatedSerializer
  import com.lightbend.lagom.scaladsl.api.deser.StrictMessageSerializer
  import com.lightbend.lagom.scaladsl.api.transport.MessageProtocol

  import scala.collection.immutable

  class ProtobufSerializer extends StrictMessageSerializer[Order] {

    private final val serializer = {
      new NegotiatedSerializer[Order, ByteString]() {
        override def protocol: MessageProtocol =
          MessageProtocol(Some("application/octet-stream"))

        def serialize(order: Order) = {
          val builder = ByteString.createBuilder
          order.writeTo(builder.asOutputStream)
          builder.result
        }
      }
    }

    private final val deserializer = {
      new NegotiatedDeserializer[Order, ByteString] {
        override def deserialize(bytes: ByteString) =
          Order.parseFrom(bytes.iterator.asInputStream)
      }
    }

    override def serializerForRequest =
      serializer
    override def deserializer(protocol: MessageProtocol) =
      deserializer
    override def serializerForResponse(
        acceptedMessageProtocols: immutable.Seq[MessageProtocol]
    ) = serializer
  }
  //#protobuf
}
